File: frontend/src/api/statistiche.ts
import { api } from "@/lib/api"
import type { EsercizioLibreria } from "@/api/schede"

export interface SerieTemporale {
  labels: string[]
  valori: number[]
}

export interface Riepilogo {
  totale_sessioni: number
  sessioni_30_giorni: number
  volume_totale_kg: number
  durata_media_min: number | null
  ultima_sessione: string | null
}

export interface Aderenza {
  sessioni: SerieTemporale
  riepilogo: { media: number; n_sessioni: number } | null
}

export interface Progressione {
  labels: string[]
  valori: number[]
  unita: string
  nome: string
}

export interface PR {
  id: number
  esercizio_libreria_id: number
  esercizio: string
  tipo: "peso" | "reps" | "tempo"
  valore: number
  peso_kg: number | null
  ripetizioni: number | null
  data: string
  sessione_id: number | null
  etichetta: string
}

export const statisticheApi = {
  riepilogo: () => api.get<Riepilogo>("/statistiche/riepilogo"),
  volume: () => api.get<SerieTemporale>("/statistiche/volume"),
  ripetizioni: () => api.get<SerieTemporale>("/statistiche/ripetizioni"),
  frequenza: () => api.get<SerieTemporale>("/statistiche/frequenza"),
  aderenza: () => api.get<Aderenza>("/statistiche/aderenza"),
  progressione: (esercizioId: number) =>
    api.get<Progressione>(`/statistiche/progressione?esercizio_id=${esercizioId}`),
  eserciziConDati: () => api.get<EsercizioLibreria[]>("/statistiche/esercizi-con-dati"),
  pr: (storico = false) => api.get<PR[]>(`/pr${storico ? "?storico=1" : ""}`),
}

File: blueprints/api/statistiche.py
"""API REST per la dashboard statistiche (sola lettura).

Split in un endpoint per grafico invece di un'unica vista aggregata: il
frontend React può così fare fetch paralleli e mostrare skeleton indipendenti
per ogni pannello, invece di aspettare tutta la pagina prima di mostrare
qualcosa. Riusa `services/stats.py` e `services/pr.py` cosi' come sono.
"""

from flask import Blueprint, request

from schemas import api_ok
from serializers import serialize_esercizio_libreria, serialize_pr
from services import stats
from services.pr import pr_attuali, storico_pr

bp = Blueprint("api_statistiche", __name__, url_prefix="/api/statistiche")
bp_pr = Blueprint("api_pr", __name__, url_prefix="/api")


@bp.get("/riepilogo")
def riepilogo_route():
    dati = stats.riepilogo()
    if dati["ultima_sessione"] is not None:
        dati["ultima_sessione"] = dati["ultima_sessione"].isoformat()
    return api_ok(dati)


@bp.get("/volume")
def volume_route():
    return api_ok(stats.volume_nel_tempo())


@bp.get("/ripetizioni")
def ripetizioni_route():
    return api_ok(stats.ripetizioni_nel_tempo())


@bp.get("/frequenza")
def frequenza_route():
    settimane = request.args.get("settimane", type=int) or 12
    return api_ok(stats.frequenza_settimanale(settimane))


@bp.get("/aderenza")
def aderenza_route():
    limite = request.args.get("limite", type=int) or 15
    return api_ok(
        {
            "sessioni": stats.aderenza_schede(limite),
            "riepilogo": stats.aderenza_riepilogo(limite),
        }
    )


@bp.get("/progressione")
def progressione_route():
    esercizio_id = request.args.get("esercizio_id", type=int)
    if not esercizio_id:
        return api_ok({"labels": [], "valori": [], "unita": "", "nome": ""})
    return api_ok(stats.progressione_esercizio(esercizio_id))


@bp.get("/esercizi-con-dati")
def esercizi_con_dati_route():
    return api_ok([serialize_esercizio_libreria(e) for e in stats.esercizi_con_dati()])


@bp_pr.get("/pr")
def pr_route():
    if request.args.get("storico"):
        limite = request.args.get("limite", type=int) or 100
        record = storico_pr()[:limite]
    else:
        record = pr_attuali()
    return api_ok([serialize_pr(p) for p in record])
